crates/saya/core/src/prover/scheduler.rs

```diff
@@ -0,0 +1,100 @@
+use futures::future::BoxFuture;
+use futures::FutureExt;
+
+use super::state_diff::ProvedStateDiff;
+use super::{prove, ProverIdentifier};
+
+type Proof = String;
+
+async fn combine_proofs(first: Proof, second: Proof) -> anyhow::Result<Proof> {
+    // TODO: Insert the real `merge program` here
+    let proof = first + " & " + &second;
+
+    tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+
+    Ok(proof)
+}
+
+// Return type based on: https://rust-lang.github.io/async-book/07_workarounds/04_recursion.html.
+pub fn prove_recursively(
+    mut inputs: Vec<ProvedStateDiff>,
+    prover: ProverIdentifier,
+) -> BoxFuture<'static, anyhow::Result<Proof>> {
+    async move {
+        if inputs.len() <= 1 {
+            prove(inputs[0].serialize(), prover).await
+        } else {
+            let last = inputs.split_off(inputs.len() / 2);
+
+            let proofs = tokio::try_join!(
+                tokio::spawn(async move { prove_recursively(inputs, prover).await }),
+                tokio::spawn(async move { prove_recursively(last, prover).await })
+            )?;
+
+            combine_proofs(proofs.0?, proofs.1?).await
```

Comment:
Do we want to extract some info here to send prepared input to the new program that will combine the proofs? Or should we extract the content we need directly from the proof inside the provable (merger) program?